core/modules/replica/IndexJob.h

```diff
@@ -0,0 +1,292 @@
+
+/**
+ * The structure IndexJobResult represents a combined result received
+ * from worker services upon a completion of the job.
+ */
+struct IndexJobResult {
+    /// MySQL-specific errors (if any) for chunks are stored in this map
+    std::map<std::string,           // worker
+             std::map<unsigned int, // chunk
+                      std::string>> error;
+};
+
+/**
+ * Class IndexJob is a base class for a family of jobs which broadcast
+ * the "secondary index" retrieval requests for the relevant chunks to
+ * the workers. Results are either dumped into the specified folder or
+ * directly loaded into the "secondary index" of a database.
+ */
+class IndexJob : public Job  {
+public:
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<IndexJob> Ptr;
+
+    /// The function type for notifications on the completion of the request
+    typedef std::function<void(Ptr)> CallbackType;
+
+    /// @return default options object for this type of a request
+    static Job::Options const& defaultOptions();
+
+    /// @return the unique name distinguishing this class from other types of jobs
+    static std::string typeName();
+
+    /// Possible destinations where the harvested data would go
+    enum Destination {
+        DISCARD,    // do nothing with the data
+        FILE,       // write all data into a file
+        FOLDER,     // write each chunk's data as a separate file at a folder
+        TABLE       // write into the specified or standard "secondary index" table
+    };
+
+    /// @return the string representation for a value of the Destination option
+    static std::string toString(Destination destination);
+
+    /// @return a value of the enumerator Destination parsed from the input string
+    /// @throw invalid_argument if the input value doen't match any known option of
+    ///        the enumerator type
+    static Destination fromString(std::string const& str);
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param database
+     *   the name of a database for which the "secondary index" is built
+     *
+     * @param hasTransactions
+     *   if 'true' then the database's "director" tables are expected to be partitioned,
+     *   and the job will extract data (including column "qserv_trans_id") from
+     *   a specific MySQL partition.
+     *
+     * @param transactionId
+     *   an identifier of a super-transaction which would limit a scope of the data
+     *   extraction requests. NOte this request will be considered only if
+     *   'hasTransactions=true'.
+     *
+     * @param allWorkers
+     *   engage all known workers regardless of their status. If the flag
+     *   is set to 'false' then only 'ENABLED' workers which are not in
+     *   the 'READ-ONLY' state will be involved into the operation.
```

Comment:
" which are not in *   the 'READ-ONLY' state" seems redundant, and confusing if more states are possible.

Reply:
Thee are two main states for workers: `ENABLED` - for workers which are allowed to participate in the Replication/Ingest operations, and `DISABLED` - for workers which are still known to the replication system, but which are temporary excluded from the operations.
And the `ENABLED` state has two sub-states: `READ-WRITE` - for forkers which can be both sources and destinations in the replication (and relevant) operations, and `READ-ONLY` - for workers which are not allowed to be used or creating new replicas. The formers may still be sources of data.
The above mentioned parameter `allWorkers` if set to `true` assumes `ENABLED & (READ-WRITE or READ-ONLY))`.